feat: update configuration doc (#1672)
AmadeusITGroup/otter: docs/configuration/CONFIGURATION_SUPPORTED_EXTRACTOR.md

> # Configuration

Reviewer: This file is missing documentation regarding the `O3rConfig` decorator

Author: https://github.com/AmadeusITGroup/otter/pull/1672/files#diff-405f767e79930a31ed8224976a6e4a8d47b9ba02aa43ec17734235e404277f0dR412

# Configuration
The component extractor is used to extract the configuration metadata from an application or a library. It extracts the list of all the config interfaces defined
in a library or an application. The output is a JSON file `component.config.metadata.json` containing an array of all the component configurations (app configurations, pages, components).
It also generates the metadata for component classes and types from an Otter library/application, outputting them in a JSON file `component.class.metadata.json`.

## How to use it

The Component Extractor is accessible via this NgCLI builder: `@o3r/components:extractor`.

> First, define your  filenames for the configuration in the `package.json` of the library/application where you run the extractor.

Reviewer 2: ```suggestion
First, define your filenames for the configuration in the `package.json` of the library/application where you run the extractor.
```

> When the extractor is run in a library, it uses this configuration to name the metadata files.

Author: comment from @cpaulve-1A:
which names are we talking about? The input or output? Names and filenames is a bit generic

https://github.com/AmadeusITGroup/otter/pull/1583/files#r1555823605

Reviewer 2: Technically, let's say it is used by the extractor to find the metadata files of a dependency to compile its content into the whole package metadata and as the path where to save the metadata file when the command is generating this file.
Not sure how to phrase it though.

Reviewer 2: It is less about application vs library than package to build vs dependencies

Reviewer 2: When the extractor is run in a project, it will use this configuration to name of the output file of the extraction. In case you want to compile the metadata of the project with its dependencies, the extractor will need to find the metadata for each dependency.
In that case, the extractor will search for this file names...

> When the extractor is run in a library, it uses this configuration to name the metadata files.

Reviewer 2: ```suggestion
The extractor uses this configuration to name its output.
```

> When the extractor is run in an application, it searches for these file names in each node_module (in the `package.json` file) of each library

Reviewer 3: Not sure about the phrasing, you need to specify the libraries (array) in the extractor configuration from the angular json, and it only goes through those libraries (not all the node modules). It checks the package json from each library, reads the fields to get the metadata file names and use those names to read the metadata file associated in each library.

# Configuration
The component extractor is used to extract the configuration metadata from an application or a library. It extracts the list of all the config interfaces defined

> in a library or an application. The output is a JSON file `component.config.metadata.json` containing an array of all the component configurations (app configurations, pages, components).

Reviewer: `component.config.metadata.json` is the default name, it can be changed

> in a library or an application. The output is a JSON file `component.config.metadata.json` containing an array of all the component configurations (app configurations, pages, components).

Reviewer: Not sure to understand the parentheses content.
Is part of the "Component Configurations"? but in this case, not sure we should indicate "app configurations". Or is the different types of configurations? If so a link to describe the different type of configs can help and maybe it would be more readable written in that way: `app, pages and components configurations` without being between parentheses

It also generates the metadata for component classes and types from an Otter library/application, outputting them in a JSON file `component.class.metadata.json`.

## How to use it

> The Component Extractor is accessible via this NgCLI builder: `@o3r/components:extractor`.

Reviewer: ```suggestion
The Component Extractor is accessible via this [Angular CLI builder](https://angular.io/guide/cli-builder): `@o3r/components:extractor`.
```

When the extractor is run in a project, it will use these file names to name the output files of the extraction. In case you want to compile the
metadata of the project with its dependencies (generally for an application that has libraries with their own metadata as well), the extractor will need to find the metadata for each dependency.
In that case, the extractor will search for these file names in the `package.json` file of each library (in the `node_modules`) 
in order to concatenate the file's metadata with the metadata of other libraries and the application's metadata.

In the `package.json` of the library:
```json
"cmsMetadata": {
  "componentFilePath": "./component.class.metadata.json",
  "configurationFilePath": "./component.config.metadata.json"
}
```

For an up-to-date documentation, run `ng help @o3r/components:extractor`

* If the __component extractor__ is run on an application with components from an Otter library, the `libraries` option can be used to concatenate the metadata files generated for the application with the ones from the specified libraries.

>   The extractor will search for the component configuration and style metadata files in the `node_modules` package of each configured library.

Reviewer: Do not explicitly speak about `node_modules`, we should prefer `dependencies` to indicate the support of non-NPM clients

  The name of the metadata files to search for is defined for each library in the `cmsMetadata` property defined in their respective `package.json` files.

> * Here is an example of an `angular.json` file:

Author: Would you suggest rephrasing or to remove the example?

Here is an example on a library:
```json
"extract-components": {
  "builder": "@o3r/components:extractor",
  "options": {
    "tsConfig": "modules/@scope/area/components/tsconfig.metadata.json",
    "configOutputFile": "modules/@scope/area/components/dist/component.config.metadata.json",
    "componentOutputFile": "modules/@scope/area/components/dist/component.class.metadata.json",
    "strictMode": true
  }
}
```

In case you have a library in a mono repository, it's important to specify both the `configOutputFile` option and the `componentOutputFile` option.
It will prevent the files generated for the library to conflict with the ones generated for the application because by default, if it is not specified, the output files will be generated at the root of the project.

For the `tsconfig`, you should include only the files that you want to parse, i.e. components and config. You can
extend your `tsconfig`, and just override the fields that you need.

Example of a `tsconfig` :

```json
{
  "extends": "./tsconfig",
  "rootDir": ".",
  "include": [
    "src/**/*.component.ts",
    "src/**/*.config.ts"
  ],
  "exclude": [

>     "Put all paths that you would like to exclude here ex : src/**/*.spec.ts"

Reviewer: ```suggestion
    // Put all paths that you would like to exclude here ex : src/**/*.spec.ts
```

(when changing the format from `json` to `json5`)

},
```

> > [!NOTE]

Author: resolved here: https://github.com/AmadeusITGroup/otter/pull/1672/files#diff-e36176a2f34e02495a4bb9da0c963e1f9bdf8ac7910b647bf9cda1875fe82899R41
